튜토리얼 작성 당시 soynlp version 은 0.0.46 입니다.

In [1]:
import sys
sys.path.insert(0, '../')

import soynlp
print(soynlp.__version__)

0.0.46


DoublespaceLineCorpus 는 텍스트 파일을 문장 별 / 문서 별로 손쉽게 읽을 수 있도록 만든 utils 입니다.

한 줄은 하나의 문서를 의미하며, 하나의 문서에서의 줄 간 구분은 '\n' 대신 두 칸 띄어쓰기 '  '로 구분합니다. 샘플 데이터는 soynlp/tutorials/doublespace_line_corpus_sample.txt 에 올려두었습니다.

In [2]:
from soynlp.utils import DoublespaceLineCorpus

corpus_fname = '../tutorials/doublespace_line_corpus_sample.txt'
corpus = DoublespaceLineCorpus(corpus_fname, iter_sent=True)

DoublespaceLineCorpus 의 iter_sent=True 를 설정하면 for loop 을 돌 때와 corpus 의 길이를 계산할 때, 문장 단위로 계산이 됩니다. 샘플 데이터에는 4065 개의 문장이 담겨있습니다.

In [3]:
print('number of sents = {}'.format(len(corpus)))

number of sents = 4065


만들어진 DoublespaceLineCorpus 의 iter_sent 를 False 로 변환하면 길이가 482 로 변합니다. 샘플 데이터는 482 개의 문서로 이뤄진 텍스트 파일이며, 4065 개의 문장으로 구성되어 있습니다.

In [4]:
corpus.iter_sent = False
print('number of docs = {}'.format(len(corpus)))

number of docs = 482


명사, 단어 추출을 위해서 iteration 을 문장 단위로 수행합니다. 이를 위해 DoublespaceLineCorpus 의 iter_sent=True 로 설정합니다.

명사 추출기의 값은 default 를 이용하였습니다. 명사 추출기에 corpus 를 입력하면 명사가 출력됩니다.

In [5]:
from soynlp.noun import LRNounExtractor_v2

corpus.iter_sent = True
noun_extractor = LRNounExtractor_v2()
nouns = noun_extractor.train_extract(corpus)

[Noun Extractor] use default predictors
[Noun Extractor] num features: pos=1260, neg=1173, common=12
[Noun Extractor] counting eojeols
[EojeolCounter] n eojeol = 17413 from 4065 sents. mem=0.080 Gb                    
[Noun Extractor] complete eojeol counter -> lr graph
[Noun Extractor] has been trained. mem=0.109 Gb
[Noun Extractor] batch prediction was completed for 6782 words
[Noun Extractor] checked compounds. discovered 242 compounds
[Noun Extractor] postprocessing detaching_features : 2651 -> 2648
[Noun Extractor] postprocessing ignore_features : 2648 -> 2628
[Noun Extractor] postprocessing ignore_NJ : 2628 -> 2482
[Noun Extractor] 2482 nouns (242 compounds) with min count=1
[Noun Extractor] flushing was done. mem=0.120 Gb                    
[Noun Extractor] 56.08 % eojeols are covered


조건을 만족하는 명사의 개수는 2482 개 입니다. 명사 추출기에서 사용자가 직접 설정할 수 있는 조건들은 noun_extractor_tutorials 를 봐주세요.

In [6]:
print('number of nouns = {}'.format(len(nouns)))

number of nouns = 2482


nouns 는 {str:collections.namedtuple} 형식의 dict 입니다. NounScore 라는 이름의 namedtuple 에는 frequency 와 score 가 저장되어 있습니다.

In [7]:
nouns['아이오아이']

NounScore(frequency=250, score=1.0)

길이가 2 이상이며, 빈도수 기준 상위 100 개인 명사를 살펴보면 아래와 같습니다.

In [8]:
selected_nouns = [(noun, score) for noun, score in nouns.items() if len(noun) > 1]
selected_nouns = sorted(selected_nouns, key=lambda x:-x[1].frequency)[:100]

for noun, score in selected_nouns:
    print('{} : {}'.format(noun, score))

트와이스 : NounScore(frequency=655, score=1.0)
데뷔 : NounScore(frequency=535, score=1.0)
기자 : NounScore(frequency=535, score=1.0)
20일 : NounScore(frequency=439, score=0.9977272727272727)
1위 : NounScore(frequency=396, score=1.0)
앨범 : NounScore(frequency=384, score=1.0)
사진 : NounScore(frequency=353, score=1.0)
방탄소년단 : NounScore(frequency=334, score=1.0)
걸그룹 : NounScore(frequency=326, score=1.0)
활동 : NounScore(frequency=325, score=0.9926470588235294)
불독 : NounScore(frequency=280, score=1.0)
팬들 : NounScore(frequency=277, score=1.0)
기록 : NounScore(frequency=277, score=1.0)
차트 : NounScore(frequency=274, score=1.0)
스트리밍 : NounScore(frequency=272, score=1.0)
컴백 : NounScore(frequency=267, score=1.0)
뮤직비디오 : NounScore(frequency=265, score=1.0)
방송 : NounScore(frequency=262, score=1.0)
아이돌 : NounScore(frequency=261, score=1.0)
출연 : NounScore(frequency=255, score=1.0)
그룹 : NounScore(frequency=253, score=1.0)
아이오아이 : NounScore(frequency=250, score=1.0)
예정 : NounScore(frequency=243, score=1.0)
공개 : NounSc